core/io/image.cpp

```diff
@@ -3165,53 +3165,52 @@ Color Image::_get_color_at_ofs(const uint8_t *ptr, uint32_t ofs) const {
 }
 
 void Image::_set_color_at_ofs(uint8_t *ptr, uint32_t ofs, const Color &p_color) {
+#define RAW_TO_8BIT(channel) CLAMP((uint8_t)Math::round(channel * 255.0), 0, 255)
 	switch (format) {
 		case FORMAT_L8: {
-			ptr[ofs] = uint8_t(CLAMP(p_color.get_v() * 255.0, 0, 255));
+			ptr[ofs] = RAW_TO_8BIT(p_color.get_v());
 		} break;
 		case FORMAT_LA8: {
-			ptr[ofs * 2 + 0] = uint8_t(CLAMP(p_color.get_v() * 255.0, 0, 255));
-			ptr[ofs * 2 + 1] = uint8_t(CLAMP(p_color.a * 255.0, 0, 255));
+			ptr[ofs * 2 + 0] = RAW_TO_8BIT(p_color.get_v());
+			ptr[ofs * 2 + 1] = RAW_TO_8BIT(p_color.a);
 		} break;
 		case FORMAT_R8: {
-			ptr[ofs] = uint8_t(CLAMP(p_color.r * 255.0, 0, 255));
+			ptr[ofs] = RAW_TO_8BIT(p_color.r);
 		} break;
 		case FORMAT_RG8: {
-			ptr[ofs * 2 + 0] = uint8_t(CLAMP(p_color.r * 255.0, 0, 255));
-			ptr[ofs * 2 + 1] = uint8_t(CLAMP(p_color.g * 255.0, 0, 255));
+			ptr[ofs * 2 + 0] = RAW_TO_8BIT(p_color.r);
+			ptr[ofs * 2 + 1] = RAW_TO_8BIT(p_color.g);
 		} break;
 		case FORMAT_RGB8: {
-			ptr[ofs * 3 + 0] = uint8_t(CLAMP(p_color.r * 255.0, 0, 255));
-			ptr[ofs * 3 + 1] = uint8_t(CLAMP(p_color.g * 255.0, 0, 255));
-			ptr[ofs * 3 + 2] = uint8_t(CLAMP(p_color.b * 255.0, 0, 255));
+			ptr[ofs * 3 + 0] = RAW_TO_8BIT(p_color.r);
+			ptr[ofs * 3 + 1] = RAW_TO_8BIT(p_color.g);
+			ptr[ofs * 3 + 2] = RAW_TO_8BIT(p_color.b);
 		} break;
 		case FORMAT_RGBA8: {
-			ptr[ofs * 4 + 0] = uint8_t(CLAMP(p_color.r * 255.0, 0, 255));
-			ptr[ofs * 4 + 1] = uint8_t(CLAMP(p_color.g * 255.0, 0, 255));
-			ptr[ofs * 4 + 2] = uint8_t(CLAMP(p_color.b * 255.0, 0, 255));
-			ptr[ofs * 4 + 3] = uint8_t(CLAMP(p_color.a * 255.0, 0, 255));
-
+			ptr[ofs * 4 + 0] = RAW_TO_8BIT(p_color.r);
+			ptr[ofs * 4 + 1] = RAW_TO_8BIT(p_color.g);
+			ptr[ofs * 4 + 2] = RAW_TO_8BIT(p_color.b);
+			ptr[ofs * 4 + 3] = RAW_TO_8BIT(p_color.a);
 		} break;
+#undef RAW_TO_8BIT
 		case FORMAT_RGBA4444: {
 			uint16_t rgba = 0;
 
-			rgba = uint16_t(CLAMP(p_color.r * 15.0, 0, 15)) << 12;
-			rgba |= uint16_t(CLAMP(p_color.g * 15.0, 0, 15)) << 8;
-			rgba |= uint16_t(CLAMP(p_color.b * 15.0, 0, 15)) << 4;
-			rgba |= uint16_t(CLAMP(p_color.a * 15.0, 0, 15));
+			rgba = CLAMP((uint16_t)Math::round(p_color.r * 15.0), 0, 15) << 12;
+			rgba |= CLAMP((uint16_t)Math::round(p_color.g * 15.0), 0, 15) << 8;
+			rgba |= CLAMP((uint16_t)Math::round(p_color.b * 15.0), 0, 15) << 4;
+			rgba |= CLAMP((uint16_t)Math::round(p_color.a * 15.0), 0, 15);
 
 			((uint16_t *)ptr)[ofs] = rgba;
-
 		} break;
 		case FORMAT_RGB565: {
 			uint16_t rgba = 0;
 
-			rgba = uint16_t(CLAMP(p_color.r * 31.0, 0, 31));
-			rgba |= uint16_t(CLAMP(p_color.g * 63.0, 0, 33)) << 5;
-			rgba |= uint16_t(CLAMP(p_color.b * 31.0, 0, 31)) << 11;
+			rgba = CLAMP((uint16_t)Math::round(p_color.r * 31.0), 0, 31);
+			rgba |= CLAMP((uint16_t)Math::round(p_color.g * 63.0), 0, 63) << 5;
```

Comment:
Using 33 was a typo here AFAICT, it should be 63 like the multiplier, which is what fits in 6 bits.

However comparing the RGB565 conversion with RGBA4444, and some sources online, I have the feel that one of the two is handling the channels backwards. Needs to sit down and write the math, or do some tests, I suspect more fixes are needed.

Reply:
> Using 33 was a typo here AFAICT, it should be 63 like the multiplier, which is what fits in 6 bits.

Yeah, definitely a typo.

> However comparing the RGB565 conversion with RGBA4444, and some sources online, I have the feel that one of the two is handling the channels backwards. Needs to sit down and write the math, or do some tests, I suspect more fixes are needed.

AFAICT it's correct for RGBA4444, and here it's doing BGR565 instead of RGB565 (BTW the variable should be renamed `rgba` -> `rgb`).

E.g. here you can compare R5G6B5 vs B5G6R5:
- https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkFormat.html#_description
- https://learn.microsoft.com/en-us/windows/win32/api/dxgiformat/ne-dxgiformat-dxgi_format#constants

`Image::_get_color_at_ofs` decodes these in the same order as they're set in here though.
https://github.com/godotengine/godot/blob/fe01776f05b1787b28b4a270d53037a3c25f4ca2/core/io/image.cpp#L3116-L3130

So if using `set_pixel`/`get_pixel` etc. it should work fine, but if creating an image from a byte buffer / obtaining such buffer from the image then the results could be unexpected for FORMAT_RGB565 as internally the buffer is assumed to be B5G6R5, not R5G6B5.

Not sure if chaning this is feasible at this point, or if it should be just documented in detail.

---

Note there's also RGBE9995 which seems to also do it backwards (compared to what its name suggests) as bitwise it's actually E5B9G9R9. But in this case there's no commonly used R9G9B9E5 format with which it could be confused (like for B5G6R5 vs R5G6B5 where both are common). So at most it's just a little confusing name.
https://github.com/godotengine/godot/blob/fe01776f05b1787b28b4a270d53037a3c25f4ca2/core/io/image.cpp#L3175-L3177
https://github.com/godotengine/godot/blob/fe01776f05b1787b28b4a270d53037a3c25f4ca2/core/math/color.h#L158